### ISSUE TITLE
HttpDo() not calculating content length for PUT

### ISSUE THREAD

**Issue by tigoe** [processing/p5.js#653]
HttpDo() appears not to calculate the content length or pass the content when you use a PUT method. May be true for other methods, I could only test PUT with my application.   A test follows:

    var payload = {};
    payload.bri = 88;
    setLight(thisLight, payload);	

    function setLight(lightNumber, data) {
      var url = 'http://example.com/
      var path = url + lightNumber + '/state';		// assemble the full URL
      var content = JSON.stringify(data);		// convert JSON obj to string

      // HttpDo below sends a request with content length of 0 and no body,
      // according to the JS console in Chrome, and WireShark:

      httpDo( path, 'PUT', content, 'json', function(response) {
        println(response);
      });

     // in contrast, the following jQuery request works perfectly:

    /*
       var request = $.ajax({
       	type: "PUT",					// use the PUT method
       	url: path,						// URL to call
       	data: content,					// body of the request
       	dataType: 'text/json'		// data type of the body
           }, function(response) {		// callback function
      		println('response: ' + response);		// server response
        });
    */

**Comment by lmccart** [comment]
the `JSON.stringify` isn't needed. in fact, the method is searching for an argument of type object to pass as the data, so it will get ignored if it's stringified.

do you have a link to a simple example of some server side code that prints the data received? i'm having a hard time confirming if it's actually working without the stringify call..

here's the line where the request is made, I can see some data is being passed, but maybe it needs to be sent in a different way.
https://github.com/processing/p5.js/blob/master/lib/p5.js#L5178

**Comment by tigoe** [comment]
Check out https://github.com/tigoe/hue-control/blob/master/client-example/client-control.js <https://github.com/tigoe/hue-control/blob/master/client-example/client-control.js>. It contacts a Hue hub, which acts as a simple HTTP server. You can see where I switched to jQuery from httpDo. Unfortunately the hub in question is not online, but there is another one, and if you mail me directly I can arrange for you to use one of the ITP hubs to see what’s going on.

**Comment by lmccart** [comment]
@tigoe did you get a chance to test this, or notice issues with it still?

**Comment by tigoe** [comment]
Yes, I can confirm it's still happening. Below is a test for GET, POST, PUT, and DELETE, with a node.js server as test.

To install the node script:
```
$ npm install express body-parser
$ node server.js
```
Then run the P5 sketch. Type A, S, D, or F for GET, POST, PUT, or DELETE, respectively. GET and POST come through fine, PUT and DELETE do not. If you run wireshark or another packet sniffer on loopback, you'll see the requests go by. At the bottom of this post are three packets, a GET, a POST, and a PUT, which shows up as OPTIONS instead. When you do a DELETE, you get a GET instead.

P5.js test sketch:

```
function setup() {

}

function keyReleased() {
  var payload = {
    foo: 88
  };
  var path = 'http://localhost:8080/';
  var content = JSON.stringify(payload);

  println("click");
  switch (key) {
    case "A": // send GET
      httpGet(path, payload, 'json', function(response) {
        println(response);
      });
      break;
    case "S": // send POST
      httpPost(path, payload, 'json', function(response) {
        println(response);
      });
      break;
    case "D": // send PUT
      httpDo(path, 'PUT', payload, 'json', function(response) {
        println(response);
      });
      break;
    case "F": // send DELETE
      httpDo(path, 'DELETE', payload, 'json', function(response) {
        println(response);
      });
      break;
  }
}
```

node.js test server:
```
var express = require('express');			// include express.js
var app = express();						// a local instance of it
var bodyParser = require('body-parser');	// include body-parser

// you need a couple of parsers for the body of a POST request:
app.use(bodyParser.json()); 						// for  application/json
app.use(bodyParser.urlencoded({ extended: true })); // for application/x-www-form-urlencoded

app.get('/', function (request, response) {
  console.log('got a GET request');
  console.log(request.query);
  response.end("Get request");
});

app.post('/', function (request, response) {
  console.log('Got a POST request');
  console.log(request.body);
  response.end("POST request");
});

app.put('/', function (request, response) {
  console.log('Got a PUT request');
  console.log(request.body);
  response.end("PUT request");
});

app.delete('/', function (request, response) {
  console.log('Got a DELETE request');
  console.log(request.body);
  response.end("DELETE request");
});

// start the server:
var server = app.listen(8080);
console.log('Server listening on port 8080');
```

packet capture:
```

No.     Time           Source                Destination           Protocol Length Info
      7 4.166754000    ::1                   ::1                   HTTP     514    GET /?foo=88 HTTP/1.1 

Frame 7: 514 bytes on wire (4112 bits), 514 bytes captured (4112 bits) on interface 0
Null/Loopback
Internet Protocol Version 6, Src: ::1 (::1), Dst: ::1 (::1)
Transmission Control Protocol, Src Port: 60728 (60728), Dst Port: 8080 (8080), Seq: 1, Ack: 1, Len: 438
Hypertext Transfer Protocol
    GET /?foo=88 HTTP/1.1\r\n
        [Expert Info (Chat/Sequence): GET /?foo=88 HTTP/1.1\r\n]
        Request Method: GET
        Request URI: /?foo=88
        Request Version: HTTP/1.1
    Host: localhost:8080\r\n
    Connection: keep-alive\r\n
    Accept: application/json, text/javascript\r\n
    Origin: http://localhost:3000\r\n
    User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.76 Safari/537.36\r\n
    Content-Type: application/x-www-form-urlencoded\r\n
    Referer: http://localhost:3000/\r\n
    Accept-Encoding: gzip, deflate\r\n
    Accept-Language: en-US,en-us;q=0.8,en;q=0.6\r\n
    \r\n
    [Full request URI: http://localhost:8080/?foo=88]
    [HTTP request 1/4]
    [Response in frame: 9]
    [Next request in frame: 11]



No.     Time           Source                Destination           Protocol Length Info
     11 5.283652000    ::1                   ::1                   HTTP     533    POST / HTTP/1.1  (application/x-www-form-urlencoded)

Frame 11: 533 bytes on wire (4264 bits), 533 bytes captured (4264 bits) on interface 0
Null/Loopback
Internet Protocol Version 6, Src: ::1 (::1), Dst: ::1 (::1)
Transmission Control Protocol, Src Port: 60728 (60728), Dst Port: 8080 (8080), Seq: 439, Ack: 135, Len: 457
Hypertext Transfer Protocol
    POST / HTTP/1.1\r\n
        [Expert Info (Chat/Sequence): POST / HTTP/1.1\r\n]
        Request Method: POST
        Request URI: /
        Request Version: HTTP/1.1
    Host: localhost:8080\r\n
    Connection: keep-alive\r\n
    Content-Length: 6\r\n
    Accept: application/json, text/javascript\r\n
    Origin: http://localhost:3000\r\n
    User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.76 Safari/537.36\r\n
    Content-Type: application/x-www-form-urlencoded\r\n
    Referer: http://localhost:3000/\r\n
    Accept-Encoding: gzip, deflate\r\n
    Accept-Language: en-US,en-us;q=0.8,en;q=0.6\r\n
    \r\n
    [Full request URI: http://localhost:8080/]
    [HTTP request 2/4]
    [Prev request in frame: 7]
    [Response in frame: 13]
    [Next request in frame: 15]
HTML Form URL Encoded: application/x-www-form-urlencoded


No.     Time           Source                Destination           Protocol Length Info
     15 6.372909000    ::1                   ::1                   HTTP     522    OPTIONS / HTTP/1.1 

Frame 15: 522 bytes on wire (4176 bits), 522 bytes captured (4176 bits) on interface 0
Null/Loopback
Internet Protocol Version 6, Src: ::1 (::1), Dst: ::1 (::1)
Transmission Control Protocol, Src Port: 60728 (60728), Dst Port: 8080 (8080), Seq: 896, Ack: 270, Len: 446
Hypertext Transfer Protocol
    OPTIONS / HTTP/1.1\r\n
        [Expert Info (Chat/Sequence): OPTIONS / HTTP/1.1\r\n]
        Request Method: OPTIONS
        Request URI: /
        Request Version: HTTP/1.1
    Host: localhost:8080\r\n
    Connection: keep-alive\r\n
    Access-Control-Request-Method: PUT\r\n
    Origin: http://localhost:3000\r\n
    User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.76 Safari/537.36\r\n
    Access-Control-Request-Headers: accept, content-type\r\n
    Accept: */*\r\n
    Referer: http://localhost:3000/\r\n
    Accept-Encoding: gzip, deflate\r\n
    Accept-Language: en-US,en-us;q=0.8,en;q=0.6\r\n
    \r\n
    [Full request URI: http://localhost:8080/]
    [HTTP request 3/4]
    [Prev request in frame: 11]
    [Response in frame: 17]
    [Next request in frame: 19]
```

**Comment by lmccart** [comment]
wonderful, thank you! i will look into this.

**Comment by sakshamsaxena** [comment]
Received GET, POST, PUT alright. DELETE shows as GET. I might fiddle around here for some time :smiley:

**Comment by tigoe** [comment]
Same results here, GET, POST, PUT working, not DELETE. Will follow the other issue. Thanks!

**Comment by tigoe** [comment]
0.5.8 seems to have fixed httpDo, but the change in signature has affected httpGet(). When I do:
`httpGet(url, callback);`

the callback doesn't get called, and I get this error in the console:
`undefined undefined  p5.min.js:7`

However, httpDo(url, 'GET', callback); works fine. And   httpGet(url, 'GET', getLights); works fine. Makes me think maybe httpGet() needs a default value for the method maybe?

**Comment by tigoe** [comment]
oh, and httpDo(url, 'PUT', callback) is working beautifully now, thanks.

**Comment by lmccart** [comment]
oh, httpGet() should not need/take a method, the GET should be implied. @limzykenneth do you want to take a look?

**Comment by limzykenneth** [comment]
Fixed with #1874 as well. It's due to an error with passing arguments from httpGet to httpDo, it affects httpPost as well.

@tigoe the `undefined undefined` error should be fixed with #1874 as well, if you are still experiencing problems then you should get a better error log, feel free to report back if needed.
